bpf/bpf_host.c

```diff
@@ -1241,28 +1264,29 @@ int to_netdev(struct __ctx_buff *ctx __maybe_unused)
 		ret = DROP_UNKNOWN_L3;
 		break;
 	}
-
 out:
 	if (IS_ERR(ret))
-		return send_drop_notify_error(ctx, srcID, ret, CTX_ACT_DROP,
-					      METRIC_EGRESS);
+		goto out_drop;
 #else
 	ret = CTX_ACT_OK;
 #endif /* ENABLE_HOST_FIREWALL */
 
-#if defined(ENABLE_NODEPORT) && \
-	(!defined(ENABLE_DSR) || \
-	 (defined(ENABLE_DSR) && defined(ENABLE_DSR_HYBRID)))
-	if ((ctx->mark & MARK_MAGIC_SNAT_DONE) != MARK_MAGIC_SNAT_DONE) {
-		ret = nodeport_nat_fwd(ctx, false);
```

Comment:
I don't think the host firewall is the main contributor to the program size here. It's only enabled where kube-proxy is disabled, which is not the case of the 4.19 pipeline yet. `nodeport_nat_fwd()` is the culprit here: the IPv4 and IPv6 paths are not in separate tail calls, which is what 26a39a8c9963325f2080ac24555067e1d17ca84b addresses.

Reply:
> I don't think the host firewall is the main contributor to the program size here. It's only enabled where kube-proxy is disabled, which is not the case of the 4.19 pipeline yet. `nodeport_nat_fwd()` is the culprit here: the IPv4 and IPv6 paths are not in separate tail calls, which is what [26a39a8](https://github.com/cilium/cilium/commit/26a39a8c9963325f2080ac24555067e1d17ca84b) addresses.

The PR here re-enables kube-proxy-replacement probing again, which is why it gets otherwise hit on 4.19. I basically was trying to figure out if I could get away with something less complex given the 1.8 release just around the corner, and this one seemed to have resolved the issue, at least when I tried locally on 4.19.57. Will see how it goes, otherwise I can also take 26a39a8 with the earlier mentioned tweaks back in.

Reply:
Took 26a39a8 back in, the 4.19 tests in CI were still 9 insns over the limit in some cases. With 26a39a8 we can also bump the NAT collision retries again which is great. I left the CILIUM_CALLS_ definitions as-is. It's not ABI of some sort so we can change it anytime, but I think we need some more investigation on what we discussed wrt populating tail call map slots. I do agree that we have an issue here which needs reworking first. E.g. if we'd move both new CALL defs at the end and leave a 'gap' in the middle, then this might not be cleaned up, if we move both into the middle, then this might be disruptive wrt datapath upgrades.